File: security/encoding.py
from __future__ import annotations

import base64
import struct
from collections.abc import Iterable


UINT16_MAX = 2**16 - 1
UINT64_MAX = 2**64 - 1
U96_MAX = 2**96 - 1


def encode_var(value: str) -> bytes:
    data = value.encode("utf-8")
    if len(data) > UINT16_MAX:
        raise ValueError("variable field exceeds 65535 bytes")
    return struct.pack(">H", len(data)) + data


def encode_uint64(value: int) -> bytes:
    if not 0 <= value <= UINT64_MAX:
        raise ValueError("uint64 value out of range")
    return value.to_bytes(8, "big")


def encode_u96(value: int) -> bytes:
    if not 0 <= value <= U96_MAX:
        raise ValueError("96-bit integer out of range")
    return value.to_bytes(12, "big")


def b64encode(data: bytes) -> str:
    return base64.b64encode(data).decode("ascii")


def b64decode(value: str) -> bytes:
    return base64.b64decode(value.encode("ascii"), validate=True)


def concat(parts: Iterable[bytes]) -> bytes:
    return b"".join(parts)

File: frontend/app/lib/secure/encoding.ts
const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();

export function utf8(value: string): Uint8Array {
  return textEncoder.encode(value);
}

export function fromUtf8(value: Uint8Array): string {
  return textDecoder.decode(value);
}

export function encodeVar(value: string): Uint8Array {
  const data = utf8(value);
  if (data.length > 0xffff) {
    throw new Error("variable field exceeds 65535 bytes");
  }
  return concat([new Uint8Array([data.length >> 8, data.length & 0xff]), data]);
}

export function encodeUint64(value: number): Uint8Array {
  return encodeBigEndian(BigInt(value), 8);
}

export function encodeU96(value: number): Uint8Array {
  return encodeBigEndian(BigInt(value), 12);
}

export function concat(parts: Uint8Array[]): Uint8Array {
  const length = parts.reduce((total, part) => total + part.length, 0);
  const output = new Uint8Array(length);
  let offset = 0;
  for (const part of parts) {
    output.set(part, offset);
    offset += part.length;
  }
  return output;
}

export function b64encode(data: Uint8Array): string {
  let binary = "";
  for (let index = 0; index < data.length; index += 1) {
    binary += String.fromCharCode(data[index]);
  }
  return btoa(binary);
}

export function b64decode(value: string): Uint8Array {
  const binary = atob(value);
  const output = new Uint8Array(binary.length);
  for (let index = 0; index < binary.length; index += 1) {
    output[index] = binary.charCodeAt(index);
  }
  return output;
}

function encodeBigEndian(value: bigint, size: number): Uint8Array {
  if (value < 0n || value >= 1n << BigInt(size * 8)) {
    throw new Error("integer out of range");
  }
  const output = new Uint8Array(size);
  let cursor = value;
  for (let index = size - 1; index >= 0; index -= 1) {
    output[index] = Number(cursor & 0xffn);
    cursor >>= 8n;
  }
  return output;
}
